sacred/exp_runner.py

```diff
@@ -0,0 +1,335 @@
+__license__ = "3-clause BSD License"  # TODO: figure out
+
+standard_library.install_aliases()
+logger = logging.getLogger(__name__)
+
+
+class Shutdown(Exception):
+    """
+    Exception for telling mongo_worker loop to quit
+    """
+
+
+class WaitQuit(Exception):
+    """
+    Exception for telling mongo_worker loop to quit
+    """
+
+
+class ReserveTimeout(Exception):
+    """No job was reserved in the alotted time
+    """
+
+
+class SacredWorker(object):
+    poll_interval = 3.0  # -- seconds
+    workdir = None
+
+    def __init__(self, observer,
+                 poll_interval=poll_interval,
+                 workdir=workdir,
+                 exp_key=None,
+                 logfilename='logfile.txt',
+                 ):
+        """
+        observer - Observer interface to jobs collection
+        poll_interval - seconds
+        workdir - string
+        exp_key - restrict reservations to this key
+        """
+        self.observer = observer
+        self.poll_interval = poll_interval
+        self.workdir = workdir
+        self.exp_key = exp_key
+        self.logfilename = logfilename
+
+    def make_log_handler(self):
+        self.log_handler = logging.FileHandler(self.logfilename)
+        self.log_handler.setFormatter(
+            logging.Formatter(
+                fmt='%(levelname)s (%(name)s): %(message)s'))
+        self.log_handler.setLevel(logging.INFO)
+
+    def run_one(self,
+                host_id=None,
+                reserve_timeout=None,
+                query=None,
+                ):
+        if host_id == None:
+            host_id = '%s:%i' % (socket.gethostname(), os.getpid()),
+
+        q = copy.copy(query or {})
+        # q['_id'] = {'$nin': sorted(blacklist)} # TODO: reinstate blacklist
+        run = None
+        start_time = time.time()
+        observer = self.observer
+        while run is None:
+            if (reserve_timeout and
+                    (time.time() - start_time) > reserve_timeout):
+                raise ReserveTimeout()
+
+            run = observer.find_queued(q)
+
+            if not run:
+                interval = (1 +
+                            numpy.random.rand() *
+                            (float(self.poll_interval) - 1.0))
+                logger.info('no job found, sleeping for %.1fs' % interval)
+                time.sleep(interval)
+        try:
+            old_status, run_id = run['status'], run['_id']
+            logger.debug('run found: %s' % str(run_id))
+
+            main_file = run['experiment']['mainfile']
+            main_module = os.path.splitext(main_file)[0]
+            logger.info('Main module: %s' % str(main_module))
+
+            mod = importlib.import_module('%s' % main_module)
+            ex = mod.ex
+
+            # ex.observers.clear()
+            # ex.observers = [o for o in ex.observers if not isinstance(o, MongoObserver)]
+            assert isinstance(ex.observers[0], MongoObserver)
```

Comment:
My idea was that the observer class could serve as the go-to experiment run manager, both for finding existing runs, overwriting them, etc. This we we could have different exp runners (or just one with different flags for other Observer types). 

I don't see the value of a queue runner without an observer, since we need to put the queued runs in some type of database regardless. What kind of usecase do you have in mind for that?

Reply:
But we could make the observerclass something purely for observing experiments, and have some type of superclass for the run database.

Reply:
I agree to all of that. I think it would be nice to extend observers to also provide a querying interface.  But this line seems to restrict the first observer of the experiment to be a MongoObserver. So if I don't add an observer in code, and only use it with `-m ...` then this assert will fail. Of course you would still want to add a MongoObserver for the exp runner.